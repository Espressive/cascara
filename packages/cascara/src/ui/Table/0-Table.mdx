---
title: Table
status: DRAFT
propTable: Table.js
---

# Public API

## data

Every object in this array will potencially be rendered as a table row.
The shape is up to you, the only requirement is that every objecct(row) must
have an attribute that uniquely identifies the object in the array.

We call this attribute `uniqueIdAttribute`, which you will be able to specify later on.

```javascript
const data = {[
  {
    id: '1',
    userEmail: 'john@doe.it',
    userName: 'John Doe',
  },
  {
    id: '2',
    userEmail: 'jane@doe.it',
    userName: 'Jane Doe',
  },
  {
    id: '3',
    userEmail: 'jim@doe.it',
    userName: 'Jim Doe',
  },
]}
```

## dataConfig

The main configuration for your table. Here you can specify the columns to display as well as the available actions (if any) for each row.
This is the core of Table's configuration, here you can define:

### actions

Actions will be appended to each row, they'll appear as buttons. Contrary to the objects in 'data', the shape of these does matter.

Actions can be of two types (for now...), the first and most common is `button`, which emits an `action` event. But if your data is `editable`, the second option
`edit` will do the trick.

```javascript
const actions: [

  // simple button
  {
    isLabeled: false,
    module: 'button',
    actionName: 'view',
    label: 'view',
    size: 'small',
  },

  // or

  // edit button combo (for editable rows)
  {
    isLabeled: false,
    module: 'edit',
    size: 'small',
    cancelLabel: 'Cancelar', // you can specify localized labels
    editLabel: 'Editar',
    saveLabel: 'Guardar',
  },
];
```

### display

Here you can describe each of the `visible columns` in your `table`. `display` items have a special shape too, good news is that you have plenty more options to choose from. Please choose the one that best fits your data type.

#### supported data types (modules)

- [checkbox](../modules/DataCheckbox)
- [email](../modules/DataEmail)
- [number](../modules/DataNumber)
- [radio](../modules/DataRadio)
- [select](../modules/DataSelect)
- [text](../modules/DataText)
- [textarea](../modules/DataTextArea)

```javasccript
/**
 * In this example we are displaying all the properties in our data objects,
 * but it's up to you how many and which properties to display. */
const display = [
  {
    attribute: 'id', // specifies the attribute in your data that represents this column
    isEditable: false, // if true, the edition mode will be activated
    isLabeled: false, // modules display a label, if false they won't
    label: 'ID', // the label to display as column header
    module: 'text', // The data module to handle the display/edit UI
  },
  {
    attribute: 'userEmail',
    isEditable: false,
    isLabeled: false,
    label: 'eMail',
    module: 'text',
  },
  {
    attribute: 'userName',
    isEditable: false,
    isLabeled: false,
    label: 'Name',
    module: 'text',
  },
];
```

## onAction (event handler)

A function you can pass to handle every event your table emits. Every event carries two pieces of information:

- target, the object that was actually clicked and
- record, the object that represents the row upon which the action was applied.

```javascript
const onAction = (target, record) => {
  switch (target.actionName) {
    case 'view':
      // do something with record
      break;

    case 'edit.save':
      // do something with record
      break;
  }
};
```

## uniqueIdAttribute

Specifies the attribute that uniquely identifies every object in the `data` array.

# Mixing everyting together

With the four props described earlier, we can easily create a table:

```jsx
<Table
  data={[
    {
      id: 1,
      userEmail: 'john@doe.it',
      userName: 'John Doe',
    },
    {
      id: 2,
      userEmail: 'jane@doe.it',
      userName: 'Jane Doe',
    },
    {
      id: 3,
      userEmail: 'jim@doe.it',
      userName: 'Jim Doe',
    },
  ]}
  dataConfig={{
    actions: [
      {
        isLabeled: false,
        module: 'button',
        actionName: 'view',
        label: 'View',
        size: 'small',
      },
    ],
    display: [
      {
        attribute: 'id',
        isEditable: false,
        isLabeled: false,
        label: 'ID',
        module: 'number',
      },
      {
        attribute: 'userEmail',
        isEditable: false,
        isLabeled: false,
        label: 'eMail',
        module: 'text',
      },
      {
        attribute: 'userName',
        isEditable: false,
        isLabeled: false,
        label: 'Name',
        module: 'text',
      },
    ],
  }}
  onAction={console.log}
  uniqueIdAttribute={'id'}
/>
```
