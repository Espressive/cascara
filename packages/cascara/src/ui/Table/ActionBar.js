import React from 'react';
import pt from 'prop-types';
// import styles from './Table.module.scss';

// external prop types
import { propTypes as actionButtonPT } from '../../modules/ActionButton';
import { propTypes as actionEditPT } from '../../modules/ActionEdit';

const propTypes = {
  actions: pt.arrayOf(
    pt.oneOfType([pt.shape(actionButtonPT), pt.shape(actionEditPT)])
  ),
  title: pt.string,
};

const ActionBar = ({ actions = [], title = '' }) => (
  <>
    {title && <span>{title}</span>}
    {actions.length && actions}
  </>
);

ActionBar.propTypes = propTypes;

export default ActionBar;
