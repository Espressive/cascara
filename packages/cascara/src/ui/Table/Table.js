import React from 'react';
import pt from 'prop-types';
import styles from './Table.module.scss';

import ErrorBoundary from '../../shared/ErrorBoundary';
import TableProvider from './context/TableProvider';

import TableHeader from './TableHeader';
import TableBody from './TableBody';

import { actionModules, dataModules } from '../../modules/ModuleKeys';

const actionModuleOptions = Object.keys(actionModules);
const dataModuleOptions = Object.keys(dataModules);

const propTypes = {
  /** An array of objects.
   *
   * Every object in this array will potencially be rendered as a table row. */
  data: pt.arrayOf(pt.shape({})),

  /** The main configuration for your table. Here you can specify the columns to display
   * as well as the available actions (if any) for each row. */
  dataConfig: pt.shape({
    /** Actions will be appended to each row, they'll appear as buttons. */
    actions: pt.arrayOf(
      pt.shape({
        module: pt.oneOf(actionModuleOptions).isRequired,
      })
    ),

    /** Here you can describe each of the visible columns in your table. */
    display: pt.arrayOf(
      pt.shape({
        module: pt.oneOf(dataModuleOptions).isRequired,
      })
    ),
  }),

  /** Event handler.
   *
   * An event handler you can pass to handle every event your table emits.*/
  onAction: pt.func,

  /** Unique ID Attribute.
   *
   * specifies the attribute that uniquely identifies every object in the 'data' array. */
  uniqueIdAttribute: pt.string,
};

/** This is a Table */
const Table = ({
  data = [],
  dataConfig = {},
  onAction = (type, data) => type,
  uniqueIdAttribute,
  ...rest
}) => {
  const { actions = [], display = [] } = dataConfig;
  let columnCount = display.length;

  if (actions.length) {
    columnCount++;
  }

  return (
    <ErrorBoundary>
      <TableProvider
        value={{
          data,
          dataConfig,
          onAction,
          uniqueIdAttribute,
        }}
        {...rest}
      >
        <table
          className={styles.Table}
          style={{
            gridTemplateColumns: `repeat(${columnCount}, auto)`,
          }}
        >
          <TableHeader />
          <TableBody />
        </table>
      </TableProvider>
    </ErrorBoundary>
  );
};

Table.propTypes = propTypes;

export default Table;
