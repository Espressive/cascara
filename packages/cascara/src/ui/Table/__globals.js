import pt from 'prop-types';
import { dataModules } from '../../modules/ModuleKeys';

const dataModuleOptions = Object.keys(dataModules);

export const PROP_TYPES = {
  // An array of action objects that will show up on the right side of a table row
  actions: pt.arrayOf(pt.shape({})),

  // An array of data objects that can potentially show up in the table.
  data: pt.arrayOf(pt.shape({})),

  // An ordered array of modules configured to show a corresponding `data` object
  dataDisplay: pt.arrayOf(
    pt.shape({
      module: pt.oneOf(dataModuleOptions).isRequired,
    })
  ),

  // The object property in a `data` object used to identify its uniqueness
  uniqueIdAttribute: pt.string,
};

export const UUID_PRIORITY_KEYS = [
  'eid',
  'uuid',
  'id',
  'sys_date_created',
  'number',
];

export const INFER_UNIQUE_ID = (objectKeys) => {
  for (const key of UUID_PRIORITY_KEYS) {
    if (objectKeys.includes(key)) {
      return key;
    }
  }
  return undefined;
};
