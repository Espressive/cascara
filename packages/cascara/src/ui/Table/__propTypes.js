import pt from 'prop-types';
import { actionModules, dataModules } from '../../modules/ModuleKeys';

const actionModuleOptions = Object.keys(actionModules);
const dataModuleOptions = Object.keys(dataModules);

export const TABLE_SHAPE = {
  /** Actions will be appended to each row, they'll appear as buttons. */
  actions: pt.shape({
    actionButtonMenuIndex: pt.number,

    modules: pt.arrayOf(
      pt.shape({
        module: pt.oneOf(actionModuleOptions).isRequired,
      })
    ),

    // Resolve record actions.
    // A function that returns the actions available to the current row
    resolveRecordActions: pt.func,
  }),

  // An array of objects.
  //
  // Every object in this array will potencially be rendered as a table row.
  data: pt.arrayOf(pt.shape({})),

  // DEPRECATED: The main configuration for your table. Here you can specify the columns to display
  // as well as the available actions (if any) for each row.
  dataConfig: pt.shape({
    /** DEPRECATED - use actions instead */
    actionButtonMenuIndex: pt.number,

    /** DEPRECATED - use actions instead */
    actions: pt.arrayOf(
      pt.shape({
        module: pt.oneOf(actionModuleOptions).isRequired,
      })
    ),

    /** DEPRECATED dataDisplay instead */
    display: pt.arrayOf(
      pt.shape({
        module: pt.oneOf(dataModuleOptions).isRequired,
      })
    ),
  }),

  /** Here you can describe each of the visible columns in your table. */
  dataDisplay: pt.arrayOf(
    pt.shape({
      module: pt.oneOf(dataModuleOptions).isRequired,
    })
  ),

  // Event handler.
  //
  // An event handler you can pass to handle every event your table emits.
  onAction: pt.func,

  // Resolve record actions.
  // A function that returns the actions available to the current row
  resolveRecordActions: pt.func,

  // Unique ID Attribute.
  //
  // specifies the attribute that uniquely identifies every object in the 'data' array.
  uniqueIdAttribute: pt.string,
};
