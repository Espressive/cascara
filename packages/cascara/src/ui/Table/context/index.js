export * from './TableContext';
export { default } from './TableContext';
