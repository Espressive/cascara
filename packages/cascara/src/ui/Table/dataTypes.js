/**
 * Data types
 * 
 * Every piece of data can be represented in different ways, 
 * i.e. a Boolean type can be appear as a Checkbox where true/false values map to Checkbox's checked/unchecked states; A Boolean can 
 * also appear as a switch with similar value -> state mapping (true -> On, false -> Off).
 * 
 * Since this multiple representation property is present in most of the supported types, and Cascara provides a flexible but slightly opinionated
 * functional design system, there's the need to define the data types and the supported representations in a clear, expressive way. 
 * 
 * An example of defining columns for a table:
 * 
 * ```js
    import dataTypes from 'dataTypes';
    
    const columns = columns: [
      {
        attribute: 'active',
        isEditable: true,
        isLabeled: false,
        label: 'Active',
        module: dataTypes.bool.asSwitch,
      },
      {
        attribute: 'eid',
        isEditable: false,
        isLabeled: false,
        label: 'ID',
        module: dataTypes.text.asText,
      },
      {
        attribute: 'email',
        isEditable: true,
        isLabeled: false,
        label: 'Email',
        module: dataTypes.text.asEmail,
      },
      ...
    ];
 * ```
 * **/
const dataTypes = {
  bool: {
    asCheckbox: 'checkbox',
    asSwitch: 'switch',
    asText: 'text',
  },
  number: {
    asEpoch: 'date',
    asNumber: 'number',
  },
  text: {
    asAvatar: 'avatar',
    asDate: 'date',
    asEmail: 'email',
    asIcon: 'icon',
    asLink: 'link',
    asRadio: 'radio',
    asSelect: 'select',
    asText: 'text',
    asTextArea: 'textarea',
  },
};

export default dataTypes;
