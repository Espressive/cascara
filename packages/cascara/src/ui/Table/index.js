import React from 'react';

import { TableContextProvider } from './context/TableContext';
import Table from './Table';

export default ({ data, dataConfig, onAction }) => (
  <TableContextProvider data={data} dataConfig={dataConfig} onAction={onAction}>
    <Table />
  </TableContextProvider>
);
