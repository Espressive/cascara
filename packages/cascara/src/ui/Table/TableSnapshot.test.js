import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'reakit';
import 'mutationobserver-shim';

import Table from '.';
import { generateFakeEmployees } from '../../lib/mock/fakeData';

// NOTE: Tests for loading, empty, and early/simple configuration state are in TableDX

describe('Table - Snapshots', () => {
  // Component tree
  // This test suite addresses the very basics of testing the Table UI.
  //
  // The first test is the snapshot, nothing special.
  //
  // Table actions
  // An extra column is appended if any or both of these are true:
  //
  // a) at least one `action` is specified in `dataConfig.actions` array
  // b) at least one column in the `dataConfig.display` array is editable
  //
  // In either cases, the extra column displays the action modules.
  //
  // The test `row actions` corresponds to condition a, whilst the test
  // `editable records` addresses condition b.
  describe('component tree', () => {
    const datasetSize = 5;
    const data = generateFakeEmployees(datasetSize);
    const actions = {
      actionButtonMenuIndex: 2,
      modules: [
        {
          content: 'view',
          isLabeled: false,
          module: 'button',
          name: 'view',
          size: 'small',
        },
        {
          content: 'delete',
          isLabeled: false,
          module: 'button',
          name: 'delete',
          size: 'small',
        },
        {
          cancelLabel: 'Cancel',
          content: 'edit',
          isLabeled: false,
          module: 'edit',
          name: 'edit',
          saveLabel: 'Save',
          size: 'small',
        },
      ],
    };

    const dataDisplay = [
      {
        attribute: 'active',
        isEditable: true,
        label: 'Active',
        module: 'checkbox',
      },
      {
        attribute: 'eid',
        isEditable: false,
        label: 'ID',
        module: 'text',
      },
      {
        attribute: 'email',
        isEditable: true,
        label: 'Email',
        module: 'email',
      },
      {
        attribute: 'country',
        isEditable: true,
        label: 'Country',
        module: 'select',
        options: [
          {
            label: 'Argentina',
            value: 'Argentina',
          },
          {
            label: 'Brazil',
            value: 'Brazil',
          },
          {
            label: 'USA',
            value: 'USA',
          },
        ],
      },
    ];

    test('snapshot test', () => {
      const view = render(
        <Provider>
          <Table
            actions={actions}
            data={data}
            dataDisplay={dataDisplay}
            uniqueIdAttribute={'eid'}
          />
        </Provider>
      ).container;

      expect(view).toMatchSnapshot();
    });

    test('without row actions', () => {
      const view = render(
        <Provider>
          <Table
            actions={actions}
            data={data}
            dataDisplay={dataDisplay}
            uniqueIdAttribute={'eid'}
          />
        </Provider>
      ).container;

      expect(view).toMatchSnapshot();
    });
  });
});
