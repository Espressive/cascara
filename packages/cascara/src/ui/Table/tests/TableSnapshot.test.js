import { render, within } from '@testing-library/react';
import 'mutationobserver-shim';

import cosmosDXFixtures from '../fixtures/TableDX.fixture';
import cosmosUXFixtures from '../fixtures/TableUX.fixture';

const {
  conditionalActions,
  withActions,
  withDeprecatedProps,
  withoutActions,
  withoutActionBar,
} = cosmosUXFixtures;

const { loading, empty, unknownActionModule, unknownDataModule } =
  cosmosDXFixtures;
// NOTE: Tests for loading, empty, and early/simple configuration state are in TableDX

describe('Table', () => {
  describe('snapshot tests', () => {
    test(`should show a loading state if the 'data' prop is 'undefined' or 'null'`, () => {
      const { baseElement } = render(loading);
      // This technically captures the entire fixture in a snapshot. We should consider if we want to only select the element within for a snapshot or not.
      // update: by default, snapshots are wrapped by a `<body>` tag, which doesn't cause snapshots to grow signigicantly
      expect(baseElement).toMatchSnapshot();
    });

    test(`should show an empty state if data is defined but has a length of 0`, () => {
      const { baseElement } = render(empty);
      expect(baseElement).toMatchSnapshot();
    });

    test('conditional actions', () => {
      const { baseElement } = render(conditionalActions);
      const [, tbody] = within(baseElement).getAllByRole('rowgroup');
      const [firstRow] = within(tbody).getAllByRole('row');

      expect(firstRow).toMatchSnapshot();
    });

    test('without actions', () => {
      const { baseElement } = render(withoutActions);

      const [, tbody] = within(baseElement).getAllByRole('rowgroup');
      const [firstRow] = within(tbody).getAllByRole('row');

      expect(firstRow).toMatchSnapshot();
    });

    test('with actions', () => {
      const { baseElement } = render(withActions);

      const [, tbody] = within(baseElement).getAllByRole('rowgroup');
      const [firstRow] = within(tbody).getAllByRole('row');

      expect(firstRow).toMatchSnapshot();
    });

    test('with deprecated props', () => {
      const { baseElement } = render(withDeprecatedProps);

      const [, tbody] = within(baseElement).getAllByRole('rowgroup');
      const [firstRow] = within(tbody).getAllByRole('row');

      expect(firstRow).toMatchSnapshot();
    });

    test('actions with non-existent module', () => {
      const { baseElement } = render(unknownActionModule);

      const [, tbody] = within(baseElement).getAllByRole('rowgroup');
      const [firstRow] = within(tbody).getAllByRole('row');

      expect(firstRow).toMatchSnapshot();
    });

    test('dataDisplay with non-existent module', () => {
      const { baseElement } = render(unknownDataModule);

      const [, tbody] = within(baseElement).getAllByRole('rowgroup');
      const [firstRow] = within(tbody).getAllByRole('row');

      expect(firstRow).toMatchSnapshot();
    });

    test('without <ActionBar />', () => {
      const { baseElement } = render(withoutActionBar);

      const [, tbody] = within(baseElement).getAllByRole('rowgroup');
      const [firstRow] = within(tbody).getAllByRole('row');

      expect(firstRow).toMatchSnapshot();
    });
  });
});
