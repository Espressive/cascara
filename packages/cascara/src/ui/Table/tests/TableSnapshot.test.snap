// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table snapshot tests actions with non-existent module 1`] = `
<tr
  class="Row"
>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="eID"
          class="Input"
        >
          1a180380-fc30-4117-ae79-dac13fc0ef38
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Name"
          class="Input"
        >
          001Ruben01_Sync
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="Protected"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="ELC"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Email"
      >
        <span
          aria-label="Created by"
          class="Input"
        >
          test@espressive.com
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Created at"
          class="Input"
        >
          2021-03-31T17:29:28.162758Z
        </span>
      </div>
    </span>
  </td>
  <td
    class="CellActions"
  >
    <button
      aria-label="Edit"
      class="ui basic icon button"
      name="edit.start"
      type="button"
    >
      Edit
    </button>
    <div
      class="Error"
      data-testid="module-error"
    >
      tequila is not a valid value for module. Try using one of [button, edit]
    </div>
    <button
      aria-controls="id-2"
      aria-expanded="false"
      aria-haspopup="menu"
      class="ui basic icon button"
      type="button"
    >
      <svg
        focusable="false"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style="transform: rotate(360deg); vertical-align: -0.143em;"
        viewBox="0 0 140 140"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M52.5 122.5a17.5 17.5 0 1 0 35 0a17.5 17.5 0 1 0-35 0zm0-52.5a17.5 17.5 0 1 0 35 0a17.5 17.5 0 1 0-35 0zm0-52.5a17.5 17.5 0 1 0 35 0a17.5 17.5 0 1 0-35 0z"
          fill="currentColor"
        />
      </svg>
    </button>
  </td>
</tr>
`;

exports[`Table snapshot tests conditional actions 1`] = `
<tr
  class="Row"
>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="eID"
          class="Input"
        >
          1a180380-fc30-4117-ae79-dac13fc0ef38
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Name"
          class="Input"
        >
          001Ruben01_Sync
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="Protected"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="ELC"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Email"
      >
        <span
          aria-label="Created by"
          class="Input"
        >
          test@espressive.com
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Created at"
          class="Input"
        >
          2021-03-31T17:29:28.162758Z
        </span>
      </div>
    </span>
  </td>
  <td
    class="CellActions"
  >
    <button
      aria-label="Edit"
      class="ui basic icon button"
      name="edit.start"
      type="button"
    >
      Edit
    </button>
  </td>
</tr>
`;

exports[`Table snapshot tests dataDisplay with non-existent module 1`] = `
<tr
  class="Row"
>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="eID"
          class="Input"
        >
          1a180380-fc30-4117-ae79-dac13fc0ef38
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Name"
          class="Input"
        >
          001Ruben01_Sync
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="Protected"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="ELC"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Email"
      >
        <span
          aria-label="Created by"
          class="Input"
        >
          test@espressive.com
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Created at"
          class="Input"
        >
          2021-03-31T17:29:28.162758Z
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <div
      class="Error"
      data-testid="module-error"
    >
      unknownModule is not a valid value for module. Try using one of [checkbox, email, json, number, radio, select, text, textarea]
    </div>
  </td>
  <td
    class="CellActions"
  >
    <button
      aria-label="Edit"
      class="ui basic icon button"
      name="edit.start"
      type="button"
    >
      Edit
    </button>
    <button
      aria-controls="id-2"
      aria-expanded="false"
      aria-haspopup="menu"
      class="ui basic icon button"
      type="button"
    >
      <svg
        focusable="false"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style="transform: rotate(360deg); vertical-align: -0.143em;"
        viewBox="0 0 140 140"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M52.5 122.5a17.5 17.5 0 1 0 35 0a17.5 17.5 0 1 0-35 0zm0-52.5a17.5 17.5 0 1 0 35 0a17.5 17.5 0 1 0-35 0zm0-52.5a17.5 17.5 0 1 0 35 0a17.5 17.5 0 1 0-35 0z"
          fill="currentColor"
        />
      </svg>
    </button>
  </td>
</tr>
`;

exports[`Table snapshot tests should show a loading state if the 'data' prop is 'undefined' or 'null' 1`] = `
<body>
  <div>
    <h3>
      Loading
    </h3>
    <p>
      The table component will not blow up if there are no values defined. It will also explicitly show a loading state if the data attribute is undefined (or null).
    </p>
    <div
      class="ui segment"
      style="height: 10em;"
    >
      <div
        class="ui active inverted dimmer"
      >
        <div
          class="ui text loader"
        >
          Loading
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Table snapshot tests should show an empty state if data is defined but has a length of 0 1`] = `
<body>
  <div>
    <h3>
      Empty
    </h3>
    <p>
      If a table gets an empty array, that should be treated different from undefined or null. An empty array should mean that there are no records and we should indicate this to users in the table itself.
    </p>
    <div
      class="ui placeholder segment"
      style="align-items: center; display: flex; height: 10em; justify-content: center;"
    >
      <em>
        No data...
      </em>
    </div>
  </div>
</body>
`;

exports[`Table snapshot tests with actions 1`] = `
<tr
  class="Row"
>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="eID"
          class="Input"
        >
          1a180380-fc30-4117-ae79-dac13fc0ef38
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Name"
          class="Input"
        >
          001Ruben01_Sync
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="Protected"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="ELC"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Email"
      >
        <span
          aria-label="Created by"
          class="Input"
        >
          test@espressive.com
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Created at"
          class="Input"
        >
          2021-03-31T17:29:28.162758Z
        </span>
      </div>
    </span>
  </td>
  <td
    class="CellActions"
  >
    <button
      aria-label="Edit"
      class="ui basic icon button"
      name="edit.start"
      type="button"
    >
      Edit
    </button>
    <button
      aria-controls="id-2"
      aria-expanded="false"
      aria-haspopup="menu"
      class="ui basic icon button"
      type="button"
    >
      <svg
        focusable="false"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style="transform: rotate(360deg); vertical-align: -0.143em;"
        viewBox="0 0 140 140"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M52.5 122.5a17.5 17.5 0 1 0 35 0a17.5 17.5 0 1 0-35 0zm0-52.5a17.5 17.5 0 1 0 35 0a17.5 17.5 0 1 0-35 0zm0-52.5a17.5 17.5 0 1 0 35 0a17.5 17.5 0 1 0-35 0z"
          fill="currentColor"
        />
      </svg>
    </button>
  </td>
</tr>
`;

exports[`Table snapshot tests with deprecated props 1`] = `
<tr
  class="Row"
>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="eID"
          class="Input"
        >
          1a180380-fc30-4117-ae79-dac13fc0ef38
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Name"
          class="Input"
        >
          001Ruben01_Sync
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="Protected"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="ELC"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Email"
      >
        <span
          aria-label="Created by"
          class="Input"
        >
          test@espressive.com
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Created at"
          class="Input"
        >
          2021-03-31T17:29:28.162758Z
        </span>
      </div>
    </span>
  </td>
  <td
    class="CellActions"
  >
    <button
      aria-label="Edit"
      class="ui basic icon button"
      name="edit.start"
      type="button"
    >
      Edit
    </button>
    <button
      aria-controls="id-2"
      aria-expanded="false"
      aria-haspopup="menu"
      class="ui basic icon button"
      type="button"
    >
      <svg
        focusable="false"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style="transform: rotate(360deg); vertical-align: -0.143em;"
        viewBox="0 0 140 140"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M52.5 122.5a17.5 17.5 0 1 0 35 0a17.5 17.5 0 1 0-35 0zm0-52.5a17.5 17.5 0 1 0 35 0a17.5 17.5 0 1 0-35 0zm0-52.5a17.5 17.5 0 1 0 35 0a17.5 17.5 0 1 0-35 0z"
          fill="currentColor"
        />
      </svg>
    </button>
  </td>
</tr>
`;

exports[`Table snapshot tests without <ActionBar /> 1`] = `
<tr
  class="Row"
>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="eID"
          class="Input"
        >
          1a180380-fc30-4117-ae79-dac13fc0ef38
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Name"
          class="Input"
        >
          001Ruben01_Sync
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="Protected"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="ELC"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Email"
      >
        <span
          aria-label="Created by"
          class="Input"
        >
          test@espressive.com
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Created at"
          class="Input"
        >
          2021-03-31T17:29:28.162758Z
        </span>
      </div>
    </span>
  </td>
  <td
    class="CellActions"
  >
    <button
      aria-label="Edit"
      class="ui basic icon button"
      name="edit.start"
      type="button"
    >
      Edit
    </button>
    <button
      class="ui basic button"
      name="viewInFAQ"
      type="button"
    >
      View
    </button>
  </td>
</tr>
`;

exports[`Table snapshot tests without actions 1`] = `
<tr
  class="Row"
>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="eID"
          class="Input"
        >
          1a180380-fc30-4117-ae79-dac13fc0ef38
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Name"
          class="Input"
        >
          001Ruben01_Sync
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="Protected"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Checkbox"
      >
        <span
          aria-label="ELC"
          class="Input"
          data-checked="true"
        />
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Email"
      >
        <span
          aria-label="Created by"
          class="Input"
        >
          test@espressive.com
        </span>
      </div>
    </span>
  </td>
  <td
    class="Cell"
  >
    <span>
      <div
        class="Text"
      >
        <span
          aria-label="Created at"
          class="Input"
        >
          2021-03-31T17:29:28.162758Z
        </span>
      </div>
    </span>
  </td>
</tr>
`;
